### ISSUE TITLE
Xml & Json PoCo parsers return valueQuantity element as SimpleQuantity instance

### ISSUE THREAD

**Issue by wmrutten** [FirelyTeam/fhir-net-api#999]
When deserializing a `StructureDefinition` that contains an `example.valueQuantity` constraint to a PoCo, both the XmlParser and the JsonParser return the `valueQuantity` constraints as a `SimpleQuantity' object instance - expecting a `Quantity` instance.

**Comment by brianpos** [comment]
This will be different for each of the versions...
So might want to be very careful with the fix and any PRs across versions.

**Comment by wmrutten** [comment]
Added unit test to demonstrate the issue:
https://github.com/FirelyTeam/fhir-net-api/blob/fix/999-ParseValueQuantity/src/Hl7.Fhir.Serialization.Tests/DeserializationTests.cs

**Comment by wmrutten** [comment]
NOTE: originally reported by a Forge STU3 user, so this issue specifically concerns the FHIR API for STU3.
Still suggest we also verify this behavior in the R4 branch, where subtypes are different.

**Comment by ewoutkramer** [comment]
This is really subtle, and the API could indeed be wrong. Let's analyse this before we move.

In R4, there's this: http://hl7.org/fhir/datatypes.html#QuantityVariations

That states that SimpleQuantity (and Money)  are *constraints*, so not datatypes, and we would never see `valueSimpleQuantity`.  However, it defines Distance, Age, Count & Duration as *datatypes*, so they could appear on the wire.

For R3 this might be different.

**Comment by ewoutkramer** [comment]
In R3, looking at http://hl7.org/fhir/STU3/datatypes.html#duration, it seems SimpleQuantity is the only "type" that is a constraint, meaning that Money is *not* a constraint.

Looking at http://hl7.org/fhir/R4/datatypes.html#open, however, we can see that for the "open" types, (in R4) suddenly Money DOES appear as a datatype, not a constraint.  

This is clearly inconsistent.

**Comment by wmrutten** [comment]
`SimpleQuantity` is defined as a constraint on `Quantity` in both STU3 and R4.

**Comment by ewoutkramer** [comment]
Yes, it seems Money is the only real inconsistency - and it's best I think for now to treat it as a datatype.

**Comment by ewoutkramer** [comment]
So, it turns out that the documentation is wrong, MoneyQuantity is indeed a constraint on Quantity, just like SimpleQuantity is.

**Comment by wmrutten** [comment]
What is the status of this issue? Can we fix this for the next release?
Note: this bug causes profile corruption in Forge when opening (deserializing) certain profiles, as reported by some users. Needs to be fixed in the API.

**Comment by ewoutkramer** [comment]
This issue just missed the 2019.3 deadline, so is now scheduled for sprint 2019.4.  This is however a major new release that might take a while.  If this is pressing we need to publish either a bugfix release (1.3.1) or batch it up and publish 1.4 if it takes too long for 2.0 to be published.

**Comment by janetr1023** [comment]
We are on R4 1.4.0 and still have this issue for doseQuantity instead of doseSimpleQuantity (although i see this fix listed in that release). I also tried the 1.5.0-beta1 prerelease and it too has the same issue. Please advise.
Thank you,
Janet

**Comment by marcovisserFurore** [comment]
Hi @janetr1023 . Thanks for your feedback. 
We will investigate this next week.

**Comment by janetr1023** [comment]
Wonderful, thank you!

**Comment by mmsmits** [comment]
Hi @janetr1023 , I'm looking into this at the moment. I I understand correctly, you expect to see a "SimpleQuantity" datatype somewhere, but you get "Quantity" right? 

Can you provide me with example code, so I can check what goes wrong?

**Comment by janetr1023** [comment]
Thanks for getting on this so quickly.

Yes the C# expects doseSimpleQuantity in the json but gets doseQuantity so the “choice” fails. 
I’m on PTO this week, so here are some snippets I could gather from my tests. Let me know if you can work with this...

Note that if you are actually going to read the MedicationStatement into a C# object, you need the Status, Subject, and Medication (Cardinality min=1). For testing just the json you can omit those and just use the Dosage.

=======
        public static MedicationStatement GenerateMedicationStatementWithDose()
        {
            var dose = new SimpleQuantity
            {
                Value = (decimal?)1.0,
                System = "http://unitsofmeasure.org",
                Code = "{tbl}"
            };

            var dosage = new Dosage
            {
                Route = MakeCodeableConcept("26643006", "Oral Route"),
                DoseAndRate = new List<Dosage.DoseAndRateComponent> { new Dosage.DoseAndRateComponent { Dose = dose } }
            };
            var ms = new MedicationStatement
            {
                // Min = 1 for Status, Subject, Medication
                Status = MedicationStatement.MedicationStatusCodes.Intended,
                Subject = new ResourceReference("Patient/123"),
                Medication = new ResourceReference("#med0309"),
                Dosage = new List<Dosage> { dosage }
            };

            return ms;
        }

=======
        [Test]
        public void MedicationStatement_Serializer_GeneratesDoseSimpleQuantity()
        {
            // Arrange
            var original = TestData.GenerateMedicationStatementWithDose();
            var serializer = SerializationHelper.CreateSerializer();

            // Act
            var json = SerializationHelper.SerializeToString(serializer, original);

            // Assert
            json.Should().Contain("doseQuantity"); // WRONG
            //json.Should().Contain("doseSimpleQuantity"); // CORRECT
        }

=======
    public static class SerializationHelper
    {
        public static JsonSerializer CreateSerializer()
        {
            var settings = new JsonSerializerSettings { Formatting = Formatting.Indented };
            FhirModelJsonSupport.ConfigureJsonSerializerSettings(settings);
            var jsonSerializer = JsonSerializer.CreateDefault(settings);
            return jsonSerializer;
        }

        public static string SerializeToString(JsonSerializer serializer, object obj)
        {
            StringBuilder sb = new StringBuilder(256);
            StringWriter sw = new StringWriter(sb, CultureInfo.InvariantCulture);
            using (JsonTextWriter jsonWriter = new JsonTextWriter(sw))
            {
                jsonWriter.Formatting = serializer.Formatting;
                serializer.Serialize(jsonWriter, obj);
            }

            return sw.ToString();
        }
    }

**Comment by mmsmits** [comment]
Thanks for sending this. I checked, and the currect behaviour is actually correct.
`SimpleQuantity` is an exception in the FHIR spec: since it's not really a datatype, but a constraint (profile) on `Quantity`. This means the serializer should still render this as `(dose)Quantity`, also when rendering the choice types.

This is also the displayed by the FHIR spec itself (http://hl7.org/fhir/dosage.html#Dosage). See `doseQuantity` and `rateQuantity` in the tree.

Here some example Json in the FHIR spec as well:
- http://hl7.org/fhir/medicationrequest0309.json.html 
- http://hl7.org/fhir/medicationrequest0316.json.html

**Comment by janetr1023** [comment]
Thank you for the detailed response!
We did see those examples, and we did create our data as doseQuantity as you suggest.
However, on binding to the C# model we continue to see the following exception: 
```
Errors:
Value is of type Hl7.Fhir.Model.Quantity, which is not an allowed choice
at Hl7.Fhir.Rest.TaskExtensions.WaitNoResult(Task task)
```
With the following from Hl7.Fhir.Model.Dosage:
```
      /// <summary>Amount of medication per dose</summary>
      [FhirElement("dose", Choice = ChoiceType.DatatypeChoice, InSummary = true, Order = 50)]
      [CLSCompliant(false)]
      [AllowedTypes(new Type[] {typeof (Range), typeof (SimpleQuantity)})]
      [DataMember]
      public Element Dose
      {
        get
        {
          return this._Dose;
        }
        set
        {
          this._Dose = value;
          this.OnPropertyChanged(nameof (Dose));
        }
      }
```

**Comment by mmsmits** [comment]
Ah thanks!

I see what you mean now. 
In our R4 release the `AllowedTypes` attribute expects `SimpleQuantity` on Dosage where it should just `Quantity`.  We should indeed fix that. Thanks for reporting and noticing this!

@brianpos I think this is an error in the T4 templates generating the datatype/resource classes. In STU3 it is still correct that Dosage.dose has an allowed datatype of `Quantity`, but this has changed to `SimpleQuanity` in R4. 
Could you check if we can fix that?

**Comment by janetr1023** [comment]
Great, thank you!

**Comment by mmsmits** [comment]
@janetr1023 I fixed it manually, and we'll try to include it in de next release.

the picture you shared indeed mentions SimpleQuantity as datatype, but if you click on SimpleQuantity you will see SimpleQuantity is merely a profile of Quantity and not a datatype. This makes SimpleQuantity an exception, that is why the deserialization is a bit weird.

I personally think we should skip SimpleQuantity in the C# datatypes as a whole, but that would be a breaking change, which we can only do when moving to version 2.0

**Comment by janetr1023** [comment]
@mmsmits Thank you for addressing this. We agree it is indeed an exception. In the meantime we are not sending the data.
We will keep an eye out of the next release and try it out. Do you have an approx timeframe for "next" and also for 2.0?

**Comment by mmsmits** [comment]
Yes next release will be somewhere this week (1.5.0-final).
Can't make any promises on 2.0, but it is work in progress.

**Comment by brianpos** [comment]
This has been a painful area in the code forever, where we've actually generated profile classes derived from the datatype class.

**Comment by janetr1023** [comment]
@mmsmits I was just checking back for "1.5.0-final" and i found 1.5.0-r4 but i don't see #999 listed in the release description: https://github.com/FirelyTeam/fhir-net-api/releases/tag/v1.5.0-r4

Is this yet available in an r4?

**Comment by mmsmits** [comment]
Yes, the fix made it to "1.5.0-final", but we forgot to mention it in the release notes. We'll update those.

**Comment by ewoutkramer** [comment]
Since this fix was done manually in one place we still have loads of SimpleQuantity members everywhere. This went unnoticed the *only* place where the type actually surfaces is a choice element - as a suffix to the serialized member name.   For non-choice types the POCO model "knows" this is actually a SimpleQuantity, but for choice types this is quite a bit harder.

And as @mmsmits and @brianpos remarks - why would we do this in the first place? Having POCO classes for constrained types brings no benefits, only pain.  So in 2.x - yes we should remove them.

See #1240.

**Comment by ewoutkramer** [comment]
Now, I've made a fix for this. Turns out this problem occurs at other places as well. This is the complete list (for R4):

Coverage (Value)
Dosage (Dose and Rate)
MedicationAdministration (Rate)
MedicationKnowledge (Characteristic, Value)
NutritionOrder (Rate)
SpecimenDefinition (MinimumVolume)

**Comment by ewoutkramer** [comment]
This is fixed in #1333 and #1339, which remove -once and for all- these classes.

**Comment by ewoutkramer** [comment]
Will be part of the upcoming 2.0 release (removing these classes is a breaking change, after all)
